odoo/models.py

```diff
@@ -5610,35 +5610,64 @@ def exists(self):
         valid_ids = set([r[0] for r in self._cr.fetchall()] + new_ids)
         return self.browse(i for i in self._ids if i in valid_ids)
 
-    def _check_recursion(self, parent=None):
+    def _check_recursion(self, field_name=None):
         """
-        Verifies that there is no loop in a hierarchical structure of records,
-        by following the parent relationship using the **parent** field until a
-        loop is detected or until a top-level record is found.
+        Verifies that there is no loop in a graph of records,
+        by following a given relationship.
+        By default the **parent** field is used as the relationship.
 
-        :param parent: optional parent field name (default: ``self._parent_name``)
+        :param field_name: optional field name (default: ``self._parent_name``)
         :return: **True** if no loop was found, **False** otherwise.
         """
-        if not parent:
-            parent = self._parent_name
-
-        # must ignore 'active' flag, ir.rules, etc. => direct SQL query
+        if not field_name:
+            field_name = self._parent_name
+        field = self._fields.get(field_name)
+        if not field:
+            raise ValueError('invalid field_name: %r' % (field_name,))
+        if not (
+            field.type in ('many2many', 'many2one')
+            and field.comodel_name == self._name
+            and field.store
+        ):
+            raise ValueError('field must be a relation on itself: %r' % (field_name,))
+
+        # must ignore 'active' flag, ir.rules, etc.
+        # => direct recursive SQL query with cycle detection
+        self.flush_model([field_name])
+        if field.type == 'many2many':
+            relation = field.relation
+            col1 = field.column1
+            col2 = field.column2
+        else:
+            relation = self._table
+            col1 = 'id'
+            col2 = field_name
         cr = self._cr
-        self.flush_model([parent])
-        for id in self.ids:
-            current_id = id
-            seen_ids = {current_id}
-            while current_id:
-                cr.execute(SQL(
-                    "SELECT %s FROM %s WHERE id = %s",
-                    SQL.identifier(parent), SQL.identifier(self._table), current_id,
-                ))
-                result = cr.fetchone()
-                current_id = result[0] if result else None
-                if current_id in seen_ids:
-                    return False
-                seen_ids.add(current_id)
-        return True
+        cr.execute(SQL(
+            """
+            WITH RECURSIVE rel AS (
+                SELECT id
+                FROM %(table)s
+                WHERE id IN %(ids)s
+            UNION ALL
+                SELECT t.%(col2)s
+                FROM rel
+                JOIN %(rel)s t
+                    ON rel.id = t.%(col1)s
+                    AND t.%(col2)s IS NOT NULL
+            ) CYCLE id SET is_cycle USING cycle_path
```

Comment:
I have rewritten it without CYCLE. In terms of performance, in the simplest case with no relations, it's a little bit slower because the query is more complex, however, in all other cases, we always send a single query to the database and don't loop in python.

```
# just a single query: much better
ps = env['res.partner'].search([], limit=20)
%timeit ps._check_recursion()
# 5.1 ms ± 207 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# -> 380 µs ± 7.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

# just a single record (it had no parent): 10% worse
p = ps[5]  # just a single one
%timeit p._check_recursion()
# 284 µs ± 5.84 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
# -> 310 µs ± 11.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

ms = env['ir.ui.menu'].search([])
len(ms)  # 283

# check all menus at one: much better
%timeit ms._check_recursion()
# 119 ms ± 1.62 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
# -> 1.2 ms ± 47.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

# check menu by menu for a better comparison: 5% better
%timeit [m._check_recursion() for m in ms]
# 122 ms ± 1.99 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
# -> 114 ms ± 5.39 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
According to my memory, the main performance problem came from m2m, especially when most nodes have multiple parents.
Coud you test it?

Reply:
I found similar results for `task.project.depend_on_ids`. You can check the performance in other cases.
In the linked PR, in the queries UNION is used which may result in slower executions than UNION ALL, there is no deduplication at each loop and we check cycles using the path.

Reply:
I push my tests in the https://github.com/odoo/odoo/pull/117806, maybe you cloud help me double check `test_107_res_partner_recursion_m2m_performance`

Reply:
The SQL implementation using the array was still faster, however it got me thinking and I wanted to test a kind of worst case. What I did: create 1000 records and generate a graph with a connectivity of ~10; generated ~9700 edges using the code below.

```python
import random
model = env['template.template']
records = model.search([], limit=1000, order='id')
check_ids = (records[0] + records[500] + records[990] + records[999]).ids
# except for a few last records, for each node create edges to random 10 following records in the list
for i, rec in enumerate(records[:-3]):
    rec.write({'rec_dep_ids': [(4, records[random.randint(i + 1, len(records) - 1)].id) for r in range(10)]})
```

Once done, I found that the previous implementation did not finish evaluating and was already very slow for record 900. So I changed the implementation to find all reachable nodes in SQL with a UNION to avoid checking paths multiple times. In the very simple case the performance is still similar to the other SQL implementation, but in bad cases such as the one above, it's the fastest. As an example `%timeit model.browse(1)._check_m2m_recursion('rec_dep_ids')` runs in ~2ms using UNION, and the python implementation took 3s and UNION ALL did not finish.